src/charm.py

```diff
@@ -251,6 +263,21 @@ def __init__(self, *args: Any) -> None:
         self.framework.observe(self.tracing.on.endpoint_changed, self._on_config_changed)
         self.framework.observe(self.tracing.on.endpoint_removed, self._on_config_changed)
 
+        self.framework.observe(
+            self.on[INTERNAL_INGRESS_RELATION_NAME].relation_joined,
+            self._configure_internal_ingress,
+        )
```

Comment:
no need to observe `relation-joined`, `relation-changed` is emitted right after it (see https://juju.is/docs/sdk/relation-name-relation-joined-event)

Reply:
it seems to be breaking the tests, not sure if it's because the harness is badly set or something else

will address this in a separate PR if it's worth fixing